### ISSUE TITLE
Can't train from scratch on CoCo

### ISSUE THREAD

**Issue by factplay1** [zylo117/Yet-Another-EfficientDet-Pytorch#122]
Hi.
Thanks for the good work.

I'm trying to train EfficientDet-d0 on CoCo from scratch (backbone is efficientnet pretrained weights). In other words, without any changes in your code, I just want to train it on CoCo (i.e. use train.py).

The code I used is this commit (one before last at the time of writing this):
[https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/85ac208de6253b1b578817550e0db03e4708d787](url)

I went through your response to other issues but didn't find solution to my problem.

I trained on 4 GPUs, with a batch of 48.
I trained for almost 50 epochs. 
Here is what I get for validation loss and mAP:
<img width="210" alt="issue-1" src="https://user-images.githubusercontent.com/50056974/79534870-a4ce7200-8030-11ea-8559-707e9a950e86.png">
<img width="421" alt="issue-2" src="https://user-images.githubusercontent.com/50056974/79534872-a5ff9f00-8030-11ea-892a-d8f5fb0eb7be.png">

It might be about to overfit. But after 50 epochs I expected to see some decent mAP.

I would very much appreciate if you could help with this.
Thank you.

**Comment by factplay1** [comment]
I have to add that I also trained from this snapshot of the code:

[https://github.com/zylo117/Yet-Another-EfficientDet-Pytorch/commit/58a1cadb2a9a6016e9414c915a3574866c5e8579](url)

which you mentioned on other issues that had some bugs.

However, with that code, I was able to achieve about 20% mAP after similar number of epochs.
Not sure if that means anything though.

**Comment by zylo117** [comment]
don't use the code before fixing loss function.

you can validate the code availability by running the tutorial.

also, can you share all of the loss graphs, not just the total? i'd like to the all of the val losses' trending.

**Comment by factplay1** [comment]
I trained after the loss function was fixed.
 The only thing not in my training code is the last commit about syncing batch norms.

Here is full log after 60+ epochs on COCO:

![issue-5](https://user-images.githubusercontent.com/50056974/79625667-5a57fe80-80df-11ea-935d-6692d9e73f64.jpg)

<img width="200" alt="issue-6" src="https://user-images.githubusercontent.com/50056974/79625670-5deb8580-80df-11ea-9af2-c62ebd917dd8.png">


<img width="281" alt="issue-8" src="https://user-images.githubusercontent.com/50056974/79625634-149b3600-80df-11ea-8cff-fc729a3ce946.png">

<img width="319" alt="issue-9" src="https://user-images.githubusercontent.com/50056974/79625636-1c5ada80-80df-11ea-88e4-cf82734e2b79.png">


<img width="192" alt="issue-7" src="https://user-images.githubusercontent.com/50056974/79625654-41e7e400-80df-11ea-9503-e40a3a7474bb.png">

I also tried several other jobs with larger batch sizes on 8 GPUs. Here is a summary of validation loss for those jobs:

![issue-4](https://user-images.githubusercontent.com/50056974/79625702-83788f00-80df-11ea-829e-3fdee7eab934.jpg)

**Comment by zylo117** [comment]
overfitting after 80k steps

**Comment by zylo117** [comment]
with bigger batchsize, you should raise your lr

**Comment by factplay1** [comment]
Yes

**Comment by factplay1** [comment]
Did you train this yourself on COCO?
With the parameters in the train script, without any changes, should I expect it to converge to a decent mAP? If so, after how many epochs? What is a good validation loss value I should expect?

**Comment by factplay1** [comment]
The other thing is that it does start to overfit, but mAP of 2% is too low...
I just want to make sure there is no other issue going on in the code...

**Comment by zylo117** [comment]
for d0, the final val reg loss is 0.035, val cls loss is 0.25, which is far lower than yours.
so you need to tune the parameters to avoid overfitting and train much longer to get better score.

**Comment by factplay1** [comment]
Thanks for the numbers!
cls loss of 0.25 doesn't sound unreachable to me.
I'll try to cope with the overfitting.
Will let you know how it'll go with some more experiments.

Thanks

**Comment by ckqsars** [comment]
I meet the same problem, when the cls loss is around 0.5 the map only 1.8%. How to solve the overfitting?

**Comment by mjanddy** [comment]
> Thanks for the numbers!
> cls loss of 0.25 doesn't sound unreachable to me.
> I'll try to cope with the overfitting.
> Will let you know how it'll go with some more experiments.
> 
> Thanks

hi,do you get the good results? Can you share the train steps?

**Comment by factplay1** [comment]
Best I have got so far is val loss of 0.33.

**Comment by ckqsars** [comment]
@factplay1 Could you show me you best mAP,  the best mAP I have got only 0.033 When the train cls loss is around 0.3

**Comment by zylo117** [comment]
> @factplay1 Could you show me you best mAP, the best mAP I have got only 0.033 When the train cls loss is around 0.3

Not training loss, but the val loss. You could be suffering overfitting.

**Comment by factplay1** [comment]
So I concluded that the code, "as is", does not reproduce COCO training from scratch (efficientdet-0).

Here is what I tried:
- Around 220+ training jobs with different configurations
- hyper-parameter tuning
- 1,2,4,8 GPU cases with various ranges of batch size
- Tried making some changes in the code so it can load pretrained efficientnet-d0 weights. This improved the loss, but still not low enough - lowest it got is 0.305, and even that gives low mAP (4%)
- Tried different kinds of optimizers, different kinds of LR scheduling techniques
- Tried 2-step training, where you train with adamw first, then try to further finetune with sgd
- Tried with the latest code as well as throughout the past 3 weeks or so
- Tried random seed vs fixed seed
- Tried different kinds of pretrained weights

I concluded that for custom datasets it works, but not for COCO training from scratch for research purposes.

**Comment by qtw1998** [comment]
> So I concluded that the code, "as is", does not reproduce COCO training from scratch (efficientdet-0).
> 
> Here is what I tried:
> 
> * Around 220+ training jobs with different configurations
> * hyper-parameter tuning
> * 1,2,4,8 GPU cases with various ranges of batch size
> * Tried making some changes in the code so it can load pretrained efficientnet-d0 weights. This improved the loss, but still not low enough - lowest it got is 0.305, and even that gives low mAP (4%)
> * Tried different kinds of optimizers, different kinds of LR scheduling techniques
> * Tried 2-step training, where you train with adamw first, then try to further finetune with sgd
> * Tried with the latest code as well as throughout the past 3 weeks or so
> * Tried random seed vs fixed seed
> * Tried different kinds of pretrained weights
> 
> I concluded that for custom datasets it works, but not for COCO training from scratch for research purposes.

why your lr is not going down?

**Comment by factplay1** [comment]
> > So I concluded that the code, "as is", does not reproduce COCO training from scratch (efficientdet-0).
> > Here is what I tried:
> > 
> > * Around 220+ training jobs with different configurations
> > * hyper-parameter tuning
> > * 1,2,4,8 GPU cases with various ranges of batch size
> > * Tried making some changes in the code so it can load pretrained efficientnet-d0 weights. This improved the loss, but still not low enough - lowest it got is 0.305, and even that gives low mAP (4%)
> > * Tried different kinds of optimizers, different kinds of LR scheduling techniques
> > * Tried 2-step training, where you train with adamw first, then try to further finetune with sgd
> > * Tried with the latest code as well as throughout the past 3 weeks or so
> > * Tried random seed vs fixed seed
> > * Tried different kinds of pretrained weights
> > 
> > I concluded that for custom datasets it works, but not for COCO training from scratch for research purposes.
> 
> why your lr is not going down?

It does go down. The default LR scheduling forces it down if loss doesn't go down after some epochs.

I also tried other kinds of LR scheduling like cosine decay - they improved, but didn't achieve good result on COCO
